ComSemApp/teacher/views.py

```diff
@@ -355,3 +357,93 @@ def delete_file(url):
         os.remove(url)
     except FileNotFoundError:
         pass
+
+
+class SpeakingPracticeAttemptReviewView(TeacherCourseViewMixin, ListView):
+    model = SpeakingPracticeAttempt
+    template_name = "ComSemApp/teacher/attempt_request_review.html"
+
+    def get_context_data(self, **kwargs: any) -> dict[str, any]:
+
+        context = super().get_context_data(**kwargs)
+
+        return context
+
+class SpeakingPracticeReviewRequestsListView(TeacherCourseViewMixin, ListView):
+    model = SpeakingPracticeAttemptReviewRequest
+    template_name = 'ComSemApp/teacher/request_list.html'
+    context_object_name = 'requests'
+
+    def get_queryset(self):
+        requests_queryset = self.course.get_speaking_practice_review_requests().order_by('review','-date')
+
+        return requests_queryset
+
+
+class SpeakingPracticeAttemptReviewCreateView(TeacherCourseViewMixin, CreateView):
+    model = SpeakingPracticeAttemptReview
+    template_name = "ComSemApp/teacher/attempt_request_info.html"
+    fields = ["request", "reviewer", "comments"]
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        attempt_id = self.kwargs.get('attempt_id')
+        attempt = get_object_or_404(SpeakingPracticeAttempt, id=attempt_id)
+        context['attempt'] = attempt
+        return context
+    
+
+    def form_invalid(self, form):
+        response = super().form_invalid(form)
+        return JsonResponse(form.errors, status=400)
+
+    def form_valid(self, form):
+        review = form.save(commit=False)
+        newScore = self.request.POST.get('newScore')
+        id = self.request.POST.get('request')
+        attempt = get_object_or_404(SpeakingPracticeAttempt, id=id)
```

Comment:
Three things:
1. You shouldn't need to get `request` from the post data, as it should be available from the form object
2. You're using the request id to look up an attempt, not an attempt id
3. This query should be inside the if block below so it doesn't get run if it doesn't need to

Reply:
request primary key is attempt so the id works for both! Since request is not a field in the form thats being created/updated it must be retrieved from the post request, not the form